api/v1alpha1/templates_common.go

```diff
@@ -73,3 +76,66 @@ type TemplateValidationStatus struct {
 	// Valid indicates whether the template passed validation or not.
 	Valid bool `json:"valid"`
 }
+
+type providersType int
+
+const (
+	bootstrapProvidersType providersType = iota
+	controlPlaneProvidersType
+	infrastructureProvidersType
+)
+
+func parseProviders[T any](providersGetter interface{ GetSpecProviders() ProvidersTupled }, typ providersType, annotations map[string]string, validationFn func(string) (T, error)) ([]ProviderTuple, error) {
+	pspec, anno := getProvidersSpecAnno(providersGetter, typ)
+	if len(pspec) > 0 {
```

Comment:
we might potentially have only part of annotations overridden in the spec(we need to inspect both annotations and spec, but prioritize values in the spec over annotations), but it's not that important for now

Reply:
I have done the part collecting the both, but the same comment: current logic counts values from the spec first, and if absent then values from the annos